File: es9/es9.py
#!/usr/bin/env python3

import os
import sys

class Marble(object):
    def __init__(self, value):
        self.value = value
        self.prev = self
        self.next = self

    def insert_after(self, marble):
        marble.next = self.next
        marble.prev = self
        self.next.prev = marble
        self.next = marble
        return marble

    def remove_and_next(self):
        self.next.prev = self.prev
        self.prev.next = self.next
        return self.next

if __name__ == "__main__":
    if len(sys.argv) < 2:
        print("Usage: {} <filename>".format(sys.argv[0]), file = sys.stderr)
        exit(-1)

    try:
        with open(sys.argv[1], "r") as f:
            ar = f.read().split(" ")
    except:
        print("Cannot open {} for reading".format(sys.argv[1]), file = sys.stderr)
        exit(-1)

    num_players = int(ar[0])
    last_marble = int(ar[6])

    players = [0 for x in range(num_players)]
    marbles = Marble(0)

    cur_marble = marbles
    cur_player = 0
    for i in range(1, last_marble+1):
        if i % 23 != 0:
            cur_marble = cur_marble.next.insert_after(Marble(i))
        else:
            for x in range(7):
                cur_marble = cur_marble.prev
            players[cur_player] += i + cur_marble.value
            cur_marble = cur_marble.remove_and_next()

        cur_player = (cur_player + 1) % num_players

    top_score = 0
    for score in players:
        if top_score < score:
            top_score = score

    print("Top score: {}".format(top_score))

File: es9/es9.go
package main

import (
	"fmt"
	"os"
)

type Marble struct {
	Value int64
	Prev  *Marble
	Next  *Marble
}

func (m *Marble) insertAfter(obj *Marble) *Marble {
	obj.Next = m.Next
	obj.Prev = m
	m.Next.Prev = obj
	m.Next = obj
	return obj
}

func (m *Marble) removeAndNext() *Marble {
	m.Next.Prev = m.Prev
	m.Prev.Next = m.Next
	return m.Next
}

func makeMarble(value int64) *Marble {
	m := Marble{Value: value}
	m.Prev = &m
	m.Next = &m
	return &m
}

func main() {
	rv := mainWithInt()
	if rv != 0 {
		os.Exit(rv)
	}
}

func mainWithInt() int {
	if len(os.Args) < 2 {
		fmt.Fprintf(os.Stderr, "Usage: %s <filename>\n", os.Args[0])
		return -1
	}

	input, err := os.Open(os.Args[1])
	if err != nil {
		fmt.Fprintf(os.Stderr, "Cannot open %s for reading\n", os.Args[1])
		return -1
	}
	defer input.Close()

	var numPlayers, lastMarble int64
	_, err = fmt.Fscanf(input, "%d players; last marble is worth %d points",
		&numPlayers, &lastMarble)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Cannot read the numplayers and lastmarble\n")
		return -1
	}

	players := make([]int64, numPlayers)
	marbles := makeMarble(0)

	curMarble := marbles
	curPlayer := int64(0)

	for i := int64(1); i < lastMarble+1; i++ {
		if i%23 != 0 {
			curMarble = curMarble.Next.insertAfter(makeMarble(i))
		} else {
			for j := 0; j < 7; j++ {
				curMarble = curMarble.Prev
			}
			players[curPlayer] += i + curMarble.Value
			curMarble = curMarble.removeAndNext()
		}
		curPlayer = (curPlayer + 1) % numPlayers
	}

	maxScore := int64(0)
	for _, score := range players {
		if maxScore < score {
			maxScore = score
		}
	}

	println("Top score:", maxScore)

	return 0
}
